contributors/design-proposals/bootstrap-discovery.md

```diff
@@ -167,11 +172,57 @@ A new well known ConfigMap will be created in the `kube-public` namespace called
+
+We extend kubeadm with a set of flags and helper commands for managing and using these tokens.
+
+### `kubeadm init` flags
+
+* `--token` If set, this injects the bootstrap token to use when initializing the cluster.  If this is unset, then a random token is created and shown to the user.  If set explicitly to the empty string then no token is generated or created.  This token is used for both discovery and TLS bootstrap by having `usage-bootstrap-signing` and `usage-bootstrap-authentication` set on the token secret.
+* `--token-ttl` If set, this sets the TTL for the lifetime of this token.  Defaults to 0 which means "forever"
+
+### `kubeadm join` flags
+
+* `--token` This sets the token for both discovery and bootstrap auth.
+* `--discovery-url` If set this will grab the cluster-info data (a kubeconfig) from a URL. Due to the sensitive nature of this data, we will only support https URLs.  This also supports `username:password@host` syntax for doing HTTP auth.
+* `--discovery-file` If set, this will load the cluster-info from a file.
+* `--discovery-token` If set, (or set via `--token`) then we will be using the token scheme described above.
+* `--tls-bootstrap-token` (not officially part of this spec) This sets the token used to temporarily authenticate to the API server in order to submit a CSR for signing.  If `--insecure-experimental-approve-all-kubelet-csrs-for-group` is set to `system:bootstrappers` then these CSRs will be approved automatically for a hands off joining flow.
+
+Only one of `--discovery-url`, `--discovery-file` or `--discovery-token` can be set.  If more than one is set then an error is surfaced and `kubeadm join` exits.  Setting `--token` counts as setting `--discovery-token`.
+
+### `kubeadm token` commands
+
+`kubeadm` provides a set of utilities for manipulating token secrets in a running server.
+
+* `kubeadm token create [token]` Creates a token server side.  With no options this'll create a token that is used for discovery and TLS bootstrap.
+  * `[token]` The actual token value (in `id.secret` form) to write in.  If unset, a random value is generated.
+  * `--usages` A list of usages.  Defaults to `signing,authentication`.
+    * If the `signing` usage is specified, the token will be used (by the BootstrapSigner controller in the KCM) to JWS-sign the ConfigMap and can then be used for discovery.
+    * If the `authentication` usage is specified, the token can be used to authenticate for TLS bootstrap.
+  * `--ttl` The TTL for this token.  This sets the expiration of the token as a duration from the current time.  This is converted into an absolute UTC time as it is written into the token secret.
+* `kubeadm token delete <token-id>|<token-id>.<token-secret>`
+  * Users can either just specify the id or the full token.  This will delete the token if it exists.
+* `kubeadm token list`
+  * List tokens in a table form listing out the `token-id.token-secret`, the TTL, the absolute expiration time and the usages.
```

Comment:
I think it might make sense to output the Secret name in the first column here actually.
Otherwise you'll have a very hard time knowing which token matches which Secret without iterating all of them